crates/commitlog/src/commit.rs

```diff
@@ -0,0 +1,234 @@
+use std::io::{self, Read, Write};
+
+use crc32c::{Crc32cReader, Crc32cWriter};
+use spacetimedb_sats::buffer::{BufReader, DecodeError};
+
+use crate::{error::ChecksumMismatch, segment::CHECKSUM_ALGORITHM_CRC32C};
+
+pub struct Header {
+    min_tx_offset: u64,
+    n: u16,
+    len: u32,
+}
+
+impl Header {
+    pub const LEN: usize = /* offset */ 8 + /* n */ 2 + /* len */  4;
+
+    /// Read [`Self::LEN`] bytes from `reader` and interpret them as the
+    /// "header" of a [`Commit`].
+    ///
+    /// Returns `None` if:
+    ///
+    /// - The reader cannot provide exactly [`Self::LEN`] bytes
+    ///
+    ///   I.e. it is at EOF
+    ///
+    /// - Or, the read bytes are all zeroes
+    ///
+    ///   This is to allow preallocation of segments.
+    ///
+    pub fn decode<R: Read>(mut reader: R) -> io::Result<Option<Self>> {
+        let mut hdr = [0; Self::LEN];
+        if let Err(e) = reader.read_exact(&mut hdr) {
+            if e.kind() == io::ErrorKind::UnexpectedEof {
+                return Ok(None);
+            }
+
+            return Err(e);
+        }
+        match &mut hdr.as_slice() {
+            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] => Ok(None),
+            buf => {
+                let min_tx_offset = buf.get_u64().map_err(decode_error)?;
+                let n = buf.get_u16().map_err(decode_error)?;
+                let len = buf.get_u32().map_err(decode_error)?;
+
+                Ok(Some(Self { min_tx_offset, n, len }))
+            }
+        }
+    }
+}
+
+/// Entry type of a [`crate::Commitlog`].
+#[derive(Debug, Default, PartialEq)]
+pub struct Commit {
+    /// The offset of the first record in this commit.
+    ///
+    /// The offset starts from zero and is counted from the beginning of the
+    /// entire log.
+    pub min_tx_offset: u64,
+    /// The number of records in the commit.
+    pub n: u16,
+    /// A buffer of all records in the commit in serialized form.
+    ///
+    /// Readers must bring their own [`crate::Decoder`] to interpret this buffer.
+    /// `n` indicates how many records the buffer contains.
+    pub records: Vec<u8>,
+}
+
+impl Commit {
+    pub const FRAMING_LEN: usize = Header::LEN + /* crc32 */ 4;
+    pub const CHECKSUM_ALGORITHM: u8 = CHECKSUM_ALGORITHM_CRC32C;
+
+    /// The largest transaction offset in this commit.
```

Comment:
This comment is wrong AFAICT. This is the min_tx_offset of the next segment, not the max_tx_offset of this segment.

Reply:
Both is true. But if its a method on commit, then it's the max offset of this commit.